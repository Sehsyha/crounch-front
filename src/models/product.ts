export interface ProductResponse {
  products: ProductApi[];
}

export interface ProductApi {
  product_name: string;
  categories: string;
  code: string;
}

export interface Product {
  barCode: string;
  name: string;
  categories: string[];
}
