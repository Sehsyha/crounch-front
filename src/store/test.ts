import { ActionTree, MutationTree, GetterTree, Module, ActionContext } from 'vuex';
import { RootState } from '.';

export type ActionCall<S, R> = (injectee: ActionContext<S, R>, payload?: any) => any;

export function createStoreModuleMock<S, A, M, G>(initialState: S, actionKeys: A, mutationKeys: M, getterKeys: G): Module<S, RootState> {
  const actions: ActionTree<S, RootState> = {};
  const mutations: MutationTree<S> = {};
  const getters: GetterTree<S, RootState> = {};

  Object.keys(actionKeys).forEach(key => actions[key] = jest.fn());
  Object.keys(mutationKeys).forEach(key => mutations[key] = jest.fn());
  Object.keys(getterKeys).forEach(key => getters[key] = jest.fn());

  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: initialState,
  }
}
