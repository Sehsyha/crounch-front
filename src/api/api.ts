import { getAPIURL } from '@/utils/environment';
import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { FetchError } from '@/utils/error';
import { AuthModule } from '@/store/AuthModule';
import { getModule } from 'vuex-module-decorators';
import router from '@/router/router';

export interface FetchOptions {
  url: string;
  data?: object;
  method: string;
}

export function doFetch<T>(options: FetchOptions): Promise<T> {
  return fetch(getUrl(options.url), getFetchOptions(options))
    .then(res => {
      if (res.status >= 200 && res.status <= 299 && res.status !== 204) {
        return res.json()
      }
      if (res.status === 204) {
        return
      }
      if (res.status === 401) {
        const authModule = getModule(AuthModule)
        authModule.logoutAction().then(() => {
          router.push({ name: 'home' })
        })
      }

      return res.json()
        .then(body => { throw new FetchError('Erreur lors de la requÃªte', res.status, body) })
        .catch()
    })
}

function getFetchOptions({ data, method }: FetchOptions): object {
  let options: any = {
    method,
    headers: {
      Authorization: localStorage.getItem(TOKEN_STORAGE_KEY),
    },
  };

  if (method !== 'GET') {
    options.body = JSON.stringify(data);
  }

  return options;
}

function getUrl(url: string): string {
  return `${getAPIURL()}/${url}`;
}
