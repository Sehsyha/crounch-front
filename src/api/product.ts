import { doFetch } from './api';
import { Product } from '@/models/product';

export function createProduct(name: string): Promise<Product> {
  return doFetch({
    url: 'products',
    method: 'POST',
    data: { name }
  });
}

export function searchProduct(name: string): Promise<Array<Product>> {
  return doFetch({
    url: 'products/search',
    method: 'POST',
    data: { name }
  })
}
