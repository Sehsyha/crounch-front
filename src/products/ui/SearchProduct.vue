<template>
  <div>
    <h3>Rechercher un produit</h3>
    <span v-if="error">{{ error }}</span>
    <input
      id="product-search"
      type="text"
      placeholder="Rechercher un produit"
      v-model="name"
      @keyup="search"
    />
    <ul id="product-search-results">
      <template v-for="product in products">
        <li
          class="product"
          :key="product.id"
          v-if="!isProductInList(product.id)"
        >
          {{ product.name }}
          <button @click="addProduct(product)">Ajouter Ã  la liste</button>
        </li>
      </template>
    </ul>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { getModule } from 'vuex-module-decorators'
import { Component } from 'vue-property-decorator'

import { ListModule } from '@/store/list/ListModule'
import { ProductApi } from '@/products/api/ProductApi'
import { ProductSearchResponse } from '@/products/api/responses'

@Component
export default class SearchProduct extends Vue {
  public listModule: ListModule = getModule(ListModule)
  public productApi = new ProductApi()

  public name: string = ""
  public error: string = ""
  public products: ProductSearchResponse[] = []

  async search() {
    if (this.name.length < 3) {
      this.products = []
      return
    }

    try {
      this.products = await this.productApi.searchProduct(this.name)
    } catch (err) {
      this.error = err.error
    }
  }

  addProduct(product: ProductSearchResponse) {
    this.$emit('add-product', product)
  }

  public isProductInList(id: string): boolean {
    return !!this.listModule.productsInSelectedList.find(productInList => productInList.id === id)
  }
}
</script>
