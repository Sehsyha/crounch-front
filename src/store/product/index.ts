import { Module } from 'vuex';

import { actions } from './actions';
import { getters } from './getters';
import { mutations } from './mutations';
import { RootState } from '..';
import { Product } from '@/models/product';

export interface ProductState {
  products: Product[];
}

export const state: ProductState = {
  products: [],
};

const namespaced: boolean = true;
export const productNamespace = { namespace: 'product' };

export const product: Module<ProductState, RootState> = {
  namespaced,
  state,
  actions,
  getters,
  mutations,
};
