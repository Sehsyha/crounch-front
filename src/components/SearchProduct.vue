<template>
  <div>
    <h3>Rechercher un produit</h3>
    <span v-if="error">{{ error }}</span>
    <input
      id="product-search"
      type="text"
      placeholder="Rechercher un produit"
      v-model="name"
      @keyup="search"
    />
    <ul id="product-search-results">
      <template v-for="product in products">
        <li class="product" :key="product.id" v-if="!isProductInList(product.id)">
          {{ product.name }}
          <button @click="addProduct(product)">Ajouter Ã  la liste</button>
        </li>
      </template>
    </ul>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component, Emit, Prop } from 'vue-property-decorator';
import { Action, Mutation } from 'vuex-class';

import { AuthActions } from '@/store/auth/keys';
import { authNamespace } from '@/store/auth';
import { searchProduct } from '@/api/product';
import { Product } from '@/models/product';
import { Events } from '@/utils/events';

@Component
export default class SearchProduct extends Vue {
  @Prop()
  public productsInList!: Product[]

  public name: string = ""
  public error: string = ""
  public products: Product[] = []

  search() {
    if (this.name.length < 3) {
      this.products = []
      return
    }
    searchProduct(this.name)
      .then(products => this.products = products)
      .catch(err => this.error = err.error)
  }

  @Emit(Events.ADD_PRODUCT)
  addProduct(product: Product) {
    return product
  }

  public isProductInList(id: string): boolean {
    return !!this.productsInList.find(productInList => productInList.id === id)
  }
}
</script>
