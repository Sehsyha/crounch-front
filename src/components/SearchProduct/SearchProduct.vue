<template>
  <div>
    <h3>Rechercher un produit</h3>
    <span v-if="error">{{ error }}</span>
    <input
      id="product-search"
      type="text"
      placeholder="Rechercher un produit"
      v-model="name"
      @keyup="search"
    />
    <ul>
      <li class="product" v-for="product in products" :key="product.id">{{ product.name }}</li>
    </ul>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { Action, Mutation } from 'vuex-class';

import { AuthActions } from '@/store/auth/keys';
import { authNamespace } from '@/store/auth';
import { ListMutations } from '../../store/list/keys';
import { listNamespace } from '../../store/list';
import { searchProduct } from '../../api/product';
import { Product } from '../../models/product';

@Component
export default class SearchProduct extends Vue {
  public name: string = ""
  public error: string = ""
  public products: Product[] = []

  search() {
    if (this.name.length < 3) {
      this.products = []
      return
    }
    searchProduct(this.name)
      .then(products => this.products = products)
      .catch(err => this.error = err.error)
  }
}
</script>
