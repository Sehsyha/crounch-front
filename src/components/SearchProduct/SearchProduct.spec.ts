import { Wrapper } from '@vue/test-utils';

import { shallowComponent } from '@/utils/test';
import SearchProduct from './SearchProduct.vue'
import { searchProduct } from '@/api/product';
import { Product } from '@/models/product';
import { Events } from '@/utils/events';

jest.mock('@/api/product')

describe('SearchProduct', () => {
  let wrapper: Wrapper<SearchProduct>;
  const NAME = 'abc'

  const PRODUCT: Product = {
    id: 'id',
    name: NAME,
  }

  beforeEach(() => {
    wrapper = shallowComponent(SearchProduct);

    (searchProduct as jest.Mock).mockResolvedValue([PRODUCT])
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('Init', () => {
    it('Should initialize name to empty string.', () => {
      expect(wrapper.vm.$data.name).toBe('');
    });

    it('Should initialize products to empty array.', () => {
      expect(wrapper.vm.$data.products).toEqual([]);
    });
  });

  describe('Search', () => {
    it('Should not call API when name length is lower than 3.', () => {
      wrapper.setData({ name: 'a' });

      expect(searchProduct as jest.Mock).not.toHaveBeenCalled()
    })

    it('Should call API when name length is greater or equal to 3.', done => {
      wrapper.setData({ name: NAME })

      wrapper.find('[type=text]').trigger('keyup')

      setTimeout(() => {
        expect(searchProduct as jest.Mock).toHaveBeenCalledWith(NAME)
        done()
      })
    })

    it('Should render found products list.', done => {
      wrapper.setData({ name: NAME })

      wrapper.find('[type=text]').trigger('keyup')

      setTimeout(() => {
        expect(wrapper).toMatchSnapshot()
        done()
      })
    })

    it('Should remove products when input is reset.', done => {
      wrapper.setData({ products: [PRODUCT] })
      wrapper.setData({ name: '' })
      wrapper.find('[type=text]').trigger('keyup')

      setTimeout(() => {
        expect(wrapper.vm.$data.products).toEqual([])
        done()
      })
    })
  })

  describe('Add product to list event', () => {
    it('Should emit event when a product is clicked.', done => {
      wrapper.setData({ products: [PRODUCT] })

      setTimeout(() => {
        wrapper.find('.product button').trigger('click')
        expect(wrapper.emitted()[Events.ADD_PRODUCT][0][0]).toEqual(PRODUCT)
        done()
      })
    })
  })
})
