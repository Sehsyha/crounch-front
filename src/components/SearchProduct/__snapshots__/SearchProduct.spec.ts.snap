// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchProduct Search Should render found products list. 1`] = `
<div>
  <h3>Rechercher un produit</h3>
  <!----> <input id="product-search" type="text" placeholder="Rechercher un produit">
  <ul>
    <li class="product">abc</li>
  </ul>
</div>
`;

exports[`SearchProduct Should render. 1`] = `
<div>
  <h3>Rechercher un produit</h3>
  <!----> <input id="product-search" type="text" placeholder="Rechercher un produit">
  <ul></ul>
</div>
`;
